code/game/objects/items/devices/dna_scanner.dm

```diff
@@ -0,0 +1,110 @@
+		to_chat(user, SPAN_WARNING("[src] battery is dead or missing."))
+		return
+	return TRUE
+
+/obj/item/device/dna_scanner/afterattack(atom/A, mob/user, proximity)
+	if(!proximity)
+		return
+	if(!can_use(user))
+		return
+
+	if(istype(A, /obj/item/modular_computer))
+		to_chat(user, "Transfering scanned mutations to \the [A].")
+		var/obj/item/modular_computer/MC = A
+		if(!MC.enabled)
+			to_chat(user, "You cannot transfer any results to \the turned off [A].")
+		else if (!MC.dna_processor)
+			to_chat(user, "You cannot transfer any results to \the [A] without DNA processor.")
+		else
+			MC.dna_processor.addMutations(scanned_mutations)
+	else
+		if(is_valid_scan_target(A) && A.simulated)
+			user.visible_message(SPAN_NOTICE("[user] runs \the [src] over \the [A]."), range = 2)
+			if(scan_sound)
+				playsound(src, scan_sound, 30)
+			if(use_delay && !do_after(user, use_delay, A))
+				to_chat(user, "You stop scanning \the [A] with \the [src].")
+				return
+			scan(A, user)
+		else
+			to_chat(user, "You cannot get any results from \the [A] with \the [src].")
+
+/obj/item/device/dna_scanner/proc/cell_check()
+	if(!cell && suitable_cell)
+		cell = new suitable_cell(src)
+
+/obj/item/device/dna_scanner/proc/cell_use_check(charge)
+	. = TRUE
+	if(!cell || !cell.checked_use(charge))
+		to_chat(usr, SPAN_WARNING("[src] battery is dead or missing."))
+		. = FALSE
+
+/obj/item/device/dna_scanner/Initialize()
+	. = ..()
+	cell_check()
+
+/obj/item/device/dna_scanner/get_cell()
+	return cell
+
+/obj/item/device/dna_scanner/handle_atom_del(atom/A)
+	..()
+	if(A == cell)
+		cell = null
+		update_icon()
+
+/obj/item/device/dna_scanner/MouseDrop(over_object)
+	if((src.loc == usr) && istype(over_object, /obj/screen/inventory/hand) && eject_item(cell, usr))
+		cell = null
+
+/obj/item/device/dna_scanner/attackby(obj/item/C, mob/living/user)
+	if(istype(C, suitable_cell) && !cell && insert_item(C, user))
+		src.cell = C
+
+/obj/item/device/dna_scanner/proc/is_valid_scan_target(atom/O)
+	return check_kognim(O) || issuperioranimal(O)
+
+/obj/item/device/dna_scanner/proc/scan(atom/A, mob/user)
+	scanned_mutations.Add(get_mutations(A, user, src))
+	to_chat(user, "You successfully scanned \the [A].")
+
+/obj/item/device/dna_scanner/proc/get_mutations(atom/target, mob/living/user, obj/scanner)
+	if(check_kognim(target))
+		var/mob/living/carbon/human/H = target
+		return H.dna.mutations
+	if(issuperioranimal(target))
+		var/mob/living/carbon/superior_animal/roach/R = target
+		if(R.mutations)
+			return R.mutations
+		else to_chat(user, "You failed to get any results from \the [target] with \the [scanner].")
+	to_chat(user, "You failed to get any results from \the [target] with \the [scanner].")
+
+/obj/item/device/dna_scanner/proc/check_kognim(atom/target)
+	if(ishuman(target))
+		var/mob/living/carbon/human/H = target
+		if(H.reagents.total_volume)
+			for(var/A in H.reagents.reagent_list)
+				var/datum/reagent/R = A
+				if(R.id == "kognim")
+					return TRUE
```

Comment:
There is a simpler and significantly less cursed way to check for a reagent. Use `reagents.has_reagent("kognim")` or something like that.